src/base/format_impl.h

```diff
@@ -0,0 +1,1114 @@
+	                        const bool localize,
+	                        T t,
+	                        Args... args) const {
+		if (arg_index == 0) {
+			format_do_impl(out, orig_index, localize, t);
+			return false;
+		}
+		// By the way, this is not a recursive call – the compiler generates a separate
+		// function for every individual template specification, and they're all inlined.
+		return format_impl(out, orig_index, arg_index - 1, localize, args...);
+	}
+
+	static inline void add_flag_if_not_present(uint8_t& flags, const Flags f, const char d) {
+		if (flags & f) {
+			throw wexception("Repeated flag '%c'", d);
+		}
+		flags |= f;
+	}
+
+	explicit Tree(const char* format_string) : format_nodes_count_(0) {
+		bool has_positional_node = false;
+		bool has_unpositional_node = false;
+
+		for (; *format_string != '\0';) {
+			// read the next character and determine the kind of node we're looking at
+			if (*format_string != '%') {
+				// plain text, read until the next % sign
+				const char* start = format_string;
+				size_t len = 0;
+				for (; *format_string != '\0' && *format_string != '%'; ++format_string, ++len) {
+				}
+				nodes_.push_back(
+				   {std::unique_ptr<AbstractNode>(new LiteralNode(std::string(start, len))), 0});
+				continue;
+			}
+			// A format node. Let's see what it says.
+			++format_string;
+			if (*format_string == '\0') {
+				throw wexception("expected format sequence after '%%', found \\0");
+			}
+			if (*format_string == '%') {
+				nodes_.push_back({std::unique_ptr<AbstractNode>(new PercentNode), 0});
+				++format_string;
+				continue;
+			}
+
+			static unsigned format_index;
+			static unsigned precision;
+			static size_t min_width;
+			static uint8_t flags;
+			static bool min_w_found;
+			precision = kInfinitePrecision;
+			min_width = 0;
+			flags = kNone;
+			min_w_found = false;
+
+			if (*format_string >= '1' && *format_string <= '9') {
+				// Read a number. This is either a format index or the min width.
+				format_index = 0;
+				do {
+					format_index *= 10;
+					format_index += (*format_string - '0');
+					++format_string;
+				} while (*format_string >= '0' && *format_string <= '9');
+				if (*format_string == '\0') {
+					throw wexception("expected format sequence after '%%%u', found \\0", format_index);
+				}
+
+				if (*format_string == '%') {
+					// A node of the form '%N%' is complete.
+					AbstractNode* node = new WildcardNode;
+					nodes_.push_back({std::unique_ptr<AbstractNode>(node), format_index});
+					if (format_index > format_nodes_count_) {
+						format_nodes_count_ = format_index;
+						format_nodes_by_index_.resize(format_nodes_count_, nullptr);
+					}
+					if (format_nodes_by_index_[format_index - 1] != nullptr) {
+						throw wexception("duplicate use of index %u", format_index);
+					}
+					if (has_unpositional_node) {
+						throw wexception("Cannot mix positional and unpositional placeholders");
+					}
+					has_positional_node = true;
```

Comment:
Could be moved to the top. (Sorry, I tried to make a suggestion, but it didn't give a readable diff.)